File: matlab/resample_particles.m
function particles= resample_particles(particles, Nmin, doresample)
%function particles= resample_particles(particles, Nmin, doresample)
%
% Resample particles if their weight variance is such that N-effective
% is less than Nmin.
%

N= length(particles);
w= zeros(1,N);
for i=1:N, w(i)= particles(i).w; end
ws= sum(w); w= w/ws;
for i=1:N, particles(i).w= particles(i).w / ws; end

[keep, Neff] = stratified_resample(w);
if Neff < Nmin & doresample==1
    particles= particles(keep);
    for i=1:N, particles(i).w= 1/N; end
end

File: cpp/resample_particles.cpp
#include "resample_particles.h"
#include "configfile.h"
#include "stratified_resample.h"
#include <iostream>

using namespace std;

void resample_particles(vector<Particle> &particles, int Nmin, int doresample) 
{
    unsigned int N = particles.size();
	assert(particles.size() == config::NPARTICLES);
    VectorXf w(N);
	w.setZero();

    unsigned i;    
    for (i=0; i<N; i++) {
        w(i) = particles[i].w();
		//cout<<"w at "<<i<<" "<<w(i)<<endl;
    }
	
	float ws = w.sum();
	
	//TODO: ws is zero!
	//cout<<"ws is "<<ws<<endl;
    for (i=0; i<N; i++) {
        w(i) = w(i)/ws;
    }
   
 
	float Neff;
	vector<int> keep;
    stratified_resample(w,keep,Neff);

	vector<Particle> old_particles = vector<Particle>(particles);
	particles.resize(keep.size());	
	
	if ((Neff < Nmin) && (doresample == 1)) {
		for(i=0; i< keep.size(); i++) {
			particles[i] = old_particles[keep[i]]; 	
		}	
		for (i=0; i<N; i++) {
			particles[i].setW(1/N);
		}
	}		
}
